File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.java
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
      List<List<Integer>> bigList = new LinkedList<>();

      Queue<TreeNode> nextLevel = new ArrayDeque<>();
      Queue<TreeNode> currentLevel = new ArrayDeque<>();

      List<Integer> list = new LinkedList<>();
      if (root == null) return bigList;

      currentLevel.add(root);

      while (!nextLevel.isEmpty() || !currentLevel.isEmpty()) {
        if (currentLevel.isEmpty()) {
          bigList.add(list);
          list = new LinkedList<>();
          currentLevel = nextLevel;
          nextLevel = new ArrayDeque<>();
        }

        TreeNode cur = currentLevel.remove();
        if (cur == null) continue;

        list.add(cur.val);
        if (cur.left != null) nextLevel.add(cur.left);
        if (cur.right != null) nextLevel.add(cur.right);
      }

      if (list != null) bigList.add(list);
      return bigList;
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.cpp

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
      std::stack<TreeNode*> pStack;
      std::stack<TreeNode*> qStack;

      findNode(root, p, pStack);
      findNode(root, q, qStack);

      while (pStack.size() != qStack.size()) {
        if (pStack.size() > qStack.size()) pStack.pop();
        else qStack.pop();
      }

      while (pStack.top()->val != qStack.top()->val) {
        pStack.pop(); qStack.pop();
      }

      return pStack.top();
    }


    bool findNode(TreeNode* root, TreeNode* p, std::stack<TreeNode*>& stack) {
      stack.push(root);
      if (!root) {stack.pop(); return false;}

      if (root->val == p->val) return true;

      if (findNode(root->left, p, stack)) return true;
      if (findNode(root->right, p, stack)) return true;
      stack.pop();
      return false;
    }

};
